### PR TITLE
GH-42136: [CI][Go][Java][JS] Use AMD64-based macOS explicitly

### PR DESCRIPTION
**Pull request by llama90**
### Rationale for this change

To support AMD64-based macOS in CI jobs. We can no longer use `macos-11` on GitHub Actions [1]. Since `macos-latest` operates based on ARM64 [2], we are changing to `macos-12` to maintain an AMD64-based workflow.

- [1] [Actions: Upcoming changes to GitHub-hosted macOS runners](https://github.blog/changelog/2024-05-20-actions-upcoming-changes-to-github-hosted-macos-runners/)
- [2] [Available-images](https://github.com/actions/runner-images/tree/main?tab=readme-ov-file#available-images)

### What changes are included in this PR?

- Updating CI jobs to support AMD64-based macOS (`macos-12`).
  - [x] Go
  - [x] Java
  - [x] JS
- Updating CI condition to trigger workflow 
  - from `github.event_name == 'push'` 
  - to `${{ !contains(github.event.pull_request.title, 'WIP') }}`

### Are these changes tested?

Yes. tested via CI.

### Are there any user-facing changes?

No.

* GitHub Issue: #42136

### CODE REVIEW

File: .github/workflows/js.yml [hunk 1]
```diff
@@ -81,9 +81,9 @@ jobs:
         run: archery docker push debian-js
 
   macos:
-    name: AMD64 macOS 11 NodeJS ${{ matrix.node }}
-    runs-on: macos-latest
-    if: github.event_name == 'push'
+    name: AMD64 macOS 12 NodeJS ${{ matrix.node }}
+    runs-on: macos-12
+    if: ${{ !contains(github.event.pull_request.title, 'WIP') }}
     timeout-minutes: 90
```
Comment (kou) [inline .github/workflows/js.yml:87]: Could you reduce timeout to 30 or something?

https://github.com/apache/arrow/actions/runs/9550814670/job/26323703330?pr=42175 took 18m 51s.

File: .github/workflows/js.yml [hunk 2]
```diff
@@ -114,7 +114,7 @@ jobs:
   windows:
     name: AMD64 Windows NodeJS ${{ matrix.node }}
     runs-on: windows-latest
-    if: github.event_name == 'push'
+    if: ${{ !contains(github.event.pull_request.title, 'WIP') }}
```
Comment (kou) [inline .github/workflows/js.yml:117]: Could you set `timeout-minutes`?

File: .github/workflows/js.yml (hunk 1 above)
Comment (llama90) [reply .github/workflows/js.yml]: Fixed it!

File: .github/workflows/js.yml (hunk 2 above)
Comment (llama90) [reply .github/workflows/js.yml:117]: Added 45m! Workflow took 30m.

**Comment by llama90** [comment]
@kou Thank you for your review! I reflected your feedback!

**Comment by kou** [comment]
@zeroshade @lidavidm @domoritz Which of "ARM64" or "AMD64" do we want to use for Go/Java/JS CI on macOS?

**Comment by llama90** [comment]
In my opinion, all kinds of languages need to support both architectures.

Once I finish this issue, I plan to propose a new issue for them. Additionally, some CIs need further enhancement (e.g., AMD64 macOS 12 with just JDK 12).

I would appreciate any feedback or suggestions on this matter.

**Review by kou — APPROVED** [review]
+1

OK. Let's discuss it on the new issue.

**Comment by kou** [comment]
#42192 is the new issue.

**Comment by vibhatha** [comment]
@github-actions crossbow submit -g java-jars

**Comment by vibhatha** [comment]
@github-actions crossbow submit -g java

**Pull request merged**